src/protocols/user_directed_commissioning/UserDirectedCommissioningServer.cpp

```diff
@@ -242,25 +242,56 @@ CHIP_ERROR IdentificationDeclaration::ReadPayload(uint8_t * udcPayload, size_t p
             mRotatingIdLen = reader.GetLength();
             err            = reader.GetBytes(mRotatingId, sizeof(mRotatingId));
             break;
-        case kAppVendorIdListTag:
+        case kTargetAppListTag:
             // app vendor list
             {
+                ChipLogError(AppServer, "TLV found an applist");
                 chip::TLV::TLVType listContainerType = chip::TLV::kTLVType_List;
                 ReturnErrorOnFailure(reader.EnterContainer(listContainerType));
 
                 while ((err = reader.Next()) == CHIP_NO_ERROR && mNumAppVendorIds < sizeof(mAppVendorIds))
                 {
                     containerTag = reader.GetTag();
                     tagNum       = static_cast<uint8_t>(chip::TLV::TagNumFromTag(containerTag));
-                    if (tagNum == kAppVendorIdTag)
+                    if (tagNum == kTargetAppTag)
                     {
-                        err = reader.Get(mAppVendorIds[mNumAppVendorIds]);
-                        mNumAppVendorIds++;
+                        ReturnErrorOnFailure(reader.EnterContainer(outerContainerType));
+                        uint16_t appVendorId  = 0;
+                        uint16_t appProductId = 0;
+
+                        while ((err = reader.Next()) == CHIP_NO_ERROR)
+                        {
+                            containerTag = reader.GetTag();
+                            tagNum       = static_cast<uint8_t>(chip::TLV::TagNumFromTag(containerTag));
+                            if (tagNum == kAppVendorIdTag)
+                            {
+                                err = reader.Get(appVendorId);
+                            }
+                            else if (tagNum == kAppProductIdTag)
+                            {
+                                err = reader.Get(appProductId);
+                                ChipLogError(AppServer, "unsupported product id tag %d", tagNum);
+                            }
+                        }
+                        if (err == CHIP_END_OF_TLV)
+                        {
+                            ChipLogProgress(AppServer, "TLV end of struct TLV");
+                            ReturnErrorOnFailure(reader.ExitContainer(outerContainerType));
+                        }
+                        if (appVendorId != 0)
+                        {
+                            mAppVendorIds[mNumAppVendorIds] = appVendorId;
+                            mNumAppVendorIds++;
```

Comment:
Isn't this an out-of-bounds write if the TLV has more than kMaxAppVendorIds things in it?

Reply:
the outer loop performs the boundary check